File: crates/koloda/src/repo/decks.rs
use rusqlite::{params, OptionalExtension};

use crate::app::db::Database;
use crate::app::error::{error_codes, throw_known_error, AppError};
use crate::app::utility::get_current_timestamp;
use crate::domain::decks::{Deck, DeleteDeckData, InsertDeckData, UpdateDeckData};
use crate::repo::algorithms::get_algorithm;
use crate::repo::templates::get_template;

fn get_deck_row(row: &rusqlite::Row<'_>) -> Result<Deck, rusqlite::Error> {
    Ok(Deck {
        id: row.get(0)?,
        title: row.get(1)?,
        algorithm_id: row.get(2)?,
        template_id: row.get(3)?,
        created_at: row.get(4)?,
        updated_at: row.get(5)?,
    })
}

pub fn get_decks(db: &Database) -> Result<Vec<Deck>, AppError> {
    throw_known_error(error_codes::DB_GET, || {
        db.with_conn(|conn| {
            let mut stmt = conn.prepare(
                r#"
                SELECT id, title, algorithm_id, template_id, created_at, updated_at
                FROM decks
                ORDER BY created_at
                "#,
            )?;

            let decks = stmt.query_map([], get_deck_row)?.collect::<Result<Vec<_>, _>>()?;

            Ok(decks)
        })
    })
}

pub fn get_decks_by_ids(db: &Database, ids: &[i64]) -> Result<Vec<Deck>, AppError> {
    throw_known_error(error_codes::DB_GET, || {
        if ids.is_empty() {
            return Ok(Vec::new());
        }

        let placeholders: Vec<String> = ids.iter().enumerate().map(|(i, _)| format!("?{}", i + 1)).collect();
        let sql = format!(
            r#"
            SELECT id, title, algorithm_id, template_id, created_at, updated_at
            FROM decks
            WHERE id IN ({})
            ORDER BY created_at
            "#,
            placeholders.join(", ")
        );

        db.with_conn(|conn| {
            let params: Vec<&dyn rusqlite::ToSql> = ids.iter().map(|id| id as &dyn rusqlite::ToSql).collect();
            let mut stmt = conn.prepare(&sql)?;
            let decks = stmt
                .query_map(params.as_slice(), get_deck_row)?
                .collect::<Result<Vec<_>, _>>()?;

            Ok(decks)
        })
    })
}

pub fn get_deck(db: &Database, id: i64) -> Result<Option<Deck>, AppError> {
    throw_known_error(error_codes::DB_GET, || {
        db.with_conn(|conn| {
            conn.query_row(
                r#"
                SELECT id, title, algorithm_id, template_id, created_at, updated_at
                FROM decks
                WHERE id = ?1
                LIMIT 1
                "#,
                params![id],
                get_deck_row,
            )
            .optional()
            .map_err(AppError::from)
        })
    })
}

pub fn add_deck(db: &Database, data: InsertDeckData) -> Result<Deck, AppError> {
    throw_known_error(error_codes::DB_ADD, || {
        data.validate()?;

        get_algorithm(db, data.algorithm_id)?.ok_or_else(|| {
            AppError::new(
                error_codes::NOT_FOUND_DECKS_ADD_ALGORITHM,
                Some(format!("Algorithm id: {}", data.algorithm_id)),
            )
        })?;
        get_template(db, data.template_id)?.ok_or_else(|| {
            AppError::new(
                error_codes::NOT_FOUND_DECKS_ADD_TEMPLATE,
                Some(format!("Template id: {}", data.template_id)),
            )
        })?;

        let now = get_current_timestamp()?;

        let id = db.with_conn(|conn| {
            conn.execute(
                r#"
                INSERT INTO decks (title, algorithm_id, template_id, created_at, updated_at)
                VALUES (?1, ?2, ?3, ?4, NULL)
                "#,
                params![data.title, data.algorithm_id, data.template_id, now],
            )?;

            Ok(conn.last_insert_rowid())
        })?;

        get_deck(db, id)?.ok_or_else(|| AppError::new(error_codes::DB_ADD, None))
    })
}

pub fn update_deck(db: &Database, data: UpdateDeckData) -> Result<Deck, AppError> {
    throw_known_error(error_codes::DB_UPDATE, || {
        data.values.validate()?;

        get_deck(db, data.id)?.ok_or_else(|| {
            AppError::new(
                error_codes::NOT_FOUND_DECKS_UPDATE_DECK,
                Some(format!("Deck id: {}", data.id)),
            )
        })?;
        get_algorithm(db, data.values.algorithm_id)?.ok_or_else(|| {
            AppError::new(
                error_codes::NOT_FOUND_DECKS_UPDATE_ALGORITHM,
                Some(format!("Algorithm id: {}", data.values.algorithm_id)),
            )
        })?;
        get_template(db, data.values.template_id)?.ok_or_else(|| {
            AppError::new(
                error_codes::NOT_FOUND_DECKS_UPDATE_TEMPLATE,
                Some(format!("Template id: {}", data.values.template_id)),
            )
        })?;

        let now = get_current_timestamp()?;

        db.with_conn(|conn| {
            conn.execute(
                r#"
                UPDATE decks
                SET
                    title = ?1,
                    algorithm_id = ?2,
                    template_id = ?3,
                    updated_at = ?4
                WHERE id = ?5
                "#,
                params![
                    data.values.title,
                    data.values.algorithm_id,
                    data.values.template_id,
                    now,
                    data.id
                ],
            )?;

            Ok(())
        })?;

        get_deck(db, data.id)?.ok_or_else(|| AppError::new(error_codes::DB_UPDATE, None))
    })
}

pub fn delete_deck(db: &Database, data: DeleteDeckData) -> Result<(), AppError> {
    throw_known_error(error_codes::DB_DELETE, || {
        db.with_conn(|conn| {
            conn.execute("DELETE FROM decks WHERE id = ?1", params![data.id])?;

            Ok(())
        })
    })
}

File: libs/db-sqlite/src/lib/decks.ts
import { AppError, mintedUuidv7, throwKnownError } from "@koloda/app";
import { deckRowSchema, updateDeckSchema } from "@koloda/srs";
import type { Deck, DeleteDeckData, InsertDeckData, UpdateDeckData } from "@koloda/srs";
import { getAlgorithm } from "./algorithms";
import { DECK_SELECT } from "./columns";
import type { DB } from "./db";
import { parseRowOrNull, parseRows } from "./parse-rows";
import { nowMs, placeholders } from "./sql";
import { getTemplate } from "./templates";

export async function getDecks(db: DB, ids?: Deck["id"][]) {
  return throwKnownError("db.get", async () => {
    if (ids && ids.length === 0) return [];

    const filter = ids?.length ? `WHERE id IN (${placeholders(ids.length)})` : "";
    const result = await db.all(`SELECT ${DECK_SELECT} FROM decks ${filter} ORDER BY created_at`, ids);
    return parseRows(deckRowSchema, result);
  });
}

export async function getDeck(db: DB, id: Deck["id"]) {
  return throwKnownError("db.get", async () => {
    const result = await db.get(`SELECT ${DECK_SELECT} FROM decks WHERE id = ? LIMIT 1`, [id]);
    return parseRowOrNull(deckRowSchema, result);
  });
}

export async function addDeck(db: DB, data: InsertDeckData) {
  return throwKnownError("db.add", async () => {
    const algorithm = await getAlgorithm(db, data.algorithmId);
    if (!algorithm) throw new AppError("not-found.decks.add.algorithm", `Algorithm id: ${data.algorithmId}`);
    const template = await getTemplate(db, data.templateId);
    if (!template) throw new AppError("not-found.decks.add.template", `Template id: ${data.templateId}`);

    const rowId = mintedUuidv7();
    await db.run(
      `INSERT INTO decks (id, title, algorithm_id, template_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?, NULL)`,
      [rowId, data.title, data.algorithmId, data.templateId, nowMs()],
    );
    const result = await getDeck(db, rowId);
    if (!result) throw new Error("no row returned");
    return result;
  });
}

export async function updateDeck(db: DB, { id, values }: UpdateDeckData) {
  return throwKnownError("db.update", async () => {
    const payload = updateDeckSchema.parse(values);

    const existing = await getDeck(db, id);
    if (!existing) throw new AppError("not-found.decks.update.deck", `Deck id: ${id}`);
    const algorithm = await getAlgorithm(db, payload.algorithmId);
    if (!algorithm) throw new AppError("not-found.decks.update.algorithm", `Algorithm id: ${payload.algorithmId}`);
    const template = await getTemplate(db, payload.templateId);
    if (!template) throw new AppError("not-found.decks.update.template", `Template id: ${payload.templateId}`);

    await db.run(`UPDATE decks SET title = ?, algorithm_id = ?, template_id = ?, updated_at = ? WHERE id = ?`, [
      payload.title,
      payload.algorithmId,
      payload.templateId,
      nowMs(),
      id,
    ]);

    const result = await getDeck(db, id);
    if (!result) throw new Error("no row returned");
    return result;
  });
}

export async function deleteDeck(db: DB, { id }: DeleteDeckData) {
  return throwKnownError("db.delete", async () => {
    await db.transaction(async (tx) => {
      await tx.run(`DELETE FROM reviews WHERE card_id IN (SELECT id FROM cards WHERE deck_id = ?)`, [id]);
      await tx.run(`DELETE FROM cards WHERE deck_id = ?`, [id]);
      await tx.run(`DELETE FROM decks WHERE id = ?`, [id]);
    });
  });
}
